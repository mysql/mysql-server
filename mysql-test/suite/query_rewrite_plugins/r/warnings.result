#
# Test of rules where the parser gives warnings about the originial
# query, pattern or replacement.
#
call mtr.add_suppression("Plugin rewriter reported: 'Some rules failed to load.'");
# Query rewrite plugin was installed.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'LOAD DATA INFILE "tab1.data" INTO TABLE test.t1',
'LOAD DATA INFILE "tab2.data" INTO TABLE test.t1' ),
( 'DO CONCAT("a", "b")', 'DO CONCAT("c", "d")' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
SELECT * FROM query_rewrite.rewrite_rules;
id	pattern	pattern_database	replacement	enabled	message
1	LOAD DATA INFILE "tab1.data" INTO TABLE test.t1	NULL	LOAD DATA INFILE "tab2.data" INTO TABLE test.t1	YES	Statement type of pattern not supported.
2	DO CONCAT("a", "b")	NULL	DO CONCAT("c", "d")	YES	Statement type of pattern not supported.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SHOW /* I rewrote it */ WARNINGS' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
# We want the note about the query being rewritten here ...
SHOW WARNINGS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
SHOW ERRORS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
GET DIAGNOSTICS CONDITION 1 @p1 = RETURNED_SQLSTATE, @p2 = MESSAGE_TEXT;
SELECT @p1;
@p1
45000
SELECT @p2;
@p2
Loading of some rule(s) failed.
# ... but not here, since we failed parsing after seeing SHOW WARNINGS.
SHOW WARNINGS and give me a parse error, please;
ERROR 42000: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'and give me a parse error, please' at line 1
DELETE FROM query_rewrite.rewrite_rules;
# Test of warnings when we rewrite the diagnostics statement.  echo #
# We disable mtr's SHOW WARNINGS here, as it would just be confusing
# when it gets rewritten without us seeing it.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SELECT "used to be SHOW WARNINGS"' ),
( 'SELECT 1', 'SHOW WARNINGS' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
SELECT 1;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
Note	1105	Query 'SELECT 1' rewritten to 'SHOW WARNINGS' by a query rewrite plugin
SHOW WARNINGS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
Note	1105	Query 'SELECT 1' rewritten to 'SHOW WARNINGS' by a query rewrite plugin
Warnings:
Warning	1620	Plugin is busy and will be uninstalled on shutdown
# Query rewrite plugin was queued for uninstalling.
