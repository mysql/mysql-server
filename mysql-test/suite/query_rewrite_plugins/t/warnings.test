--echo #
--echo # Test of rules where the parser gives warnings about the originial
--echo # query, pattern or replacement.
--echo #

call mtr.add_suppression("Plugin rewriter reported: 'Some rules failed to load.'");

--source suite/query_rewrite_plugins/include/have_plugin_rewriter.inc
--source suite/query_rewrite_plugins/include/install_rewriter.inc

# Still many unsupported statements
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'LOAD DATA INFILE "tab1.data" INTO TABLE test.t1',
         'LOAD DATA INFILE "tab2.data" INTO TABLE test.t1' ),
       ( 'DO CONCAT("a", "b")', 'DO CONCAT("c", "d")' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();
SELECT * FROM query_rewrite.rewrite_rules;

INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SHOW /* I rewrote it */ WARNINGS' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();

--echo # We want the note about the query being rewritten here ...
SHOW WARNINGS;
SHOW ERRORS;
GET DIAGNOSTICS CONDITION 1 @p1 = RETURNED_SQLSTATE, @p2 = MESSAGE_TEXT;
SELECT @p1;
SELECT @p2;

--echo # ... but not here, since we failed parsing after seeing SHOW WARNINGS.
--error ER_PARSE_ERROR
SHOW WARNINGS and give me a parse error, please;

DELETE FROM query_rewrite.rewrite_rules;

--echo # Test of warnings when we rewrite the diagnostics statement.  echo #
--echo # We disable mtr's SHOW WARNINGS here, as it would just be confusing
--echo # when it gets rewritten without us seeing it.
--disable_warnings
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SELECT "used to be SHOW WARNINGS"' ),
       ( 'SELECT 1', 'SHOW WARNINGS' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();

SELECT 1;

SHOW WARNINGS;

--enable_warnings
--source suite/query_rewrite_plugins/include/uninstall_rewriter.inc
