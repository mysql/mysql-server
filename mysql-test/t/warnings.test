#
# Test some warnings
# sql_mode set to NULL to test warnings for LOAD DATA INFILE below
--disable_warnings
--disable_result_log
set @@session.sql_mode='';
--enable_result_log

drop table if exists t1, t2;
--enable_warnings
SET SQL_WARNINGS=1;

create table t1 (a int);
--error 1050
create table t1 (a int);
show count(*) errors;
show errors;
show warnings;
--error 1115
create table t2(a int) default charset qwerty;
show count(*) errors;
show errors;
--error 1064
create table t (i);
show count(*) errors;
show errors;
insert into t1 values (1);
insert ignore into t1 values ("hej");
insert ignore into t1 values ("hej"),("då");
set SQL_WARNINGS=1;
insert ignore into t1 values ("hej");
insert ignore into t1 values ("hej"),("då");
drop table t1;
set SQL_WARNINGS=0;

#
# Test other warnings
#

drop temporary table if exists not_exists;
drop table if exists not_exists_table;
show warnings limit 1;
drop database if exists not_exists_db;
show count(*) warnings;
create table t1(id int);
create table if not exists t1(id int);
--disable_ps_protocol
select @@warning_count;
--enable_ps_protocol
drop table t1;

#
# Test warnings for LOAD DATA INFILE
#
create table t1(a tinyint, b int not null, c date, d char(5));
load data infile '../../std_data/warnings_loaddata.dat' into table t1 fields terminated by ',';
# PS doesn't work good with @@warning_count
--disable_ps_protocol
select @@warning_count;
--enable_ps_protocol
drop table t1;

#
# Warnings from basic INSERT, UPDATE and ALTER commands
#

create table t1(a tinyint NOT NULL, b tinyint unsigned, c char(5));
insert ignore into t1 values(NULL,100,'mysql'),(10,-1,'mysql ab'),(500,256,'open source'),(20,NULL,'test');
SET sql_mode = 'NO_ENGINE_SUBSTITUTION';
alter table t1 modify c char(4);
SET sql_mode = default;
alter table t1 add d char(2);
update ignore t1 set a=NULL where a=10;
update ignore t1 set c='mysql ab' where c='test';
update ignore t1 set d=c;
create table t2(a tinyint NOT NULL, b char(3));
insert ignore into t2 select b,c from t1;
insert ignore into t2(b) values('mysqlab');
set sql_warnings=1;
insert ignore into t2(b) values('mysqlab');
set sql_warnings=0;
drop table t1, t2;

#
# Test for max_error_count
#

create table t1(a char(10));
let $1=50;
disable_query_log;
while ($1)
{
  eval insert into t1 values('mysql ab');
  dec $1;
}
enable_query_log;
alter table t1 add b char;
set max_error_count=10;
update ignore t1 set b=a;
--disable_ps_protocol
select @@warning_count;
--enable_ps_protocol

# Bug#9072
set max_error_count=0;
show variables like 'max_error_count';
update ignore t1 set b='hi';
--disable_ps_protocol
select @@warning_count;
--enable_ps_protocol
show warnings;
set max_error_count=65535;
show variables like 'max_error_count';
set max_error_count=10;
show variables like 'max_error_count';

drop table t1;

#
# Tests for show warnings limit a, b
#
create table t1 (a int);
insert into t1 (a) values (1), (2), (3), (4), (5), (6), (7), (8), (9), (10);
# should generate 10 warnings
update ignore t1 set a='abc';
show warnings limit 2, 1;
show warnings limit 0, 10;
show warnings limit 9, 1;
show warnings limit 10, 1;
show warnings limit 9, 2;
show warnings limit 0, 0;
show warnings limit 1;
show warnings limit 0;
show warnings limit 1, 0;
# make sure behaviour is consistent with select ... limit
select * from t1 limit 0;
select * from t1 limit 1, 0;
select * from t1 limit 0, 0;
drop table t1;

--echo End of 4.1 tests

#
# Bug#20778: strange characters in warning message 1366 when called in SP
#

CREATE TABLE t1( f1 CHAR(20) );
CREATE TABLE t2( f1 CHAR(20), f2 CHAR(25) );
CREATE TABLE t3( f1 CHAR(20), f2 CHAR(25), f3 DATE );

INSERT INTO t1 VALUES ( 'a`' );
INSERT INTO t2 VALUES ( 'a`', 'a`' );
INSERT INTO t3 VALUES ( 'a`', 'a`', '1000-01-1' );

DROP PROCEDURE IF EXISTS sp1;
DROP PROCEDURE IF EXISTS sp2;
DROP PROCEDURE IF EXISTS sp3;
SET sql_mode = 'NO_ENGINE_SUBSTITUTION';
delimiter //;
CREATE PROCEDURE sp1()
BEGIN
   DECLARE x NUMERIC ZEROFILL;
   SELECT f1 INTO x FROM t1 LIMIT 1;
END//
CREATE PROCEDURE sp2()
BEGIN
   DECLARE x NUMERIC ZEROFILL;
   SELECT f1 INTO x FROM t2 LIMIT 1;
END//
CREATE PROCEDURE sp3()
BEGIN
   DECLARE x NUMERIC ZEROFILL;
   SELECT f1 INTO x FROM t3 LIMIT 1;
END//
delimiter ;//
CALL sp1();
CALL sp2();
CALL sp3();

DROP PROCEDURE IF EXISTS sp1;
delimiter //;
CREATE PROCEDURE sp1()
BEGIN
declare x numeric unsigned zerofill;
SELECT f1 into x from t2 limit 1;
END//
delimiter ;//
CALL sp1();
DROP TABLE t1;
DROP TABLE t2;
DROP TABLE t3;
DROP PROCEDURE sp1;
DROP PROCEDURE sp2;
DROP PROCEDURE sp3;
SET sql_mode = default;

#
# Bug#30059: End-space truncation warnings are inconsistent or incorrect
#

create table t1 (c_char char(255), c_varchar varchar(255), c_tinytext tinytext);
create table t2 (c_tinyblob tinyblob); # not affected by bug, for regression testing
set @c = repeat(' ', 256);
set @q = repeat('q', 256);

set sql_mode = '';

insert into t1 values(@c, @c, @c);
insert into t2 values(@c);
insert into t1 values(@q, @q, @q);
insert into t2 values(@q);

set sql_mode = 'traditional';

insert into t1 values(@c, @c, @c);
--error 1406
insert into t2 values(@c);
--error 1406
insert into t1 values(@q, NULL, NULL);
--error 1406
insert into t1 values(NULL, @q, NULL);
--error 1406
insert into t1 values(NULL, NULL, @q);
--error 1406
insert into t2 values(@q);

drop table t1, t2;

#
# Bug#42364 SHOW ERRORS returns empty resultset after dropping non existent table
#
--error ER_BAD_TABLE_ERROR
DROP TABLE t1;
SHOW ERRORS;

--echo End of 5.0 tests

#
# Bug#55847: SHOW WARNINGS returns empty result set when SQLEXCEPTION is active
#

--echo
--echo -- Bug#55847
--echo

--disable_warnings
DROP TABLE IF EXISTS t1;
DROP FUNCTION IF EXISTS f1;
--enable_warnings

CREATE TABLE t1(a INT UNIQUE);

delimiter |;

CREATE FUNCTION f1(x INT) RETURNS INT
BEGIN
  INSERT INTO t1 VALUES(x);
  INSERT INTO t1 VALUES(x);
  RETURN x;
END|

delimiter ;|

--echo

--error ER_DUP_ENTRY
SHOW TABLES WHERE f1(11) = 11;

--echo

SHOW WARNINGS;

--echo

DROP TABLE t1;
DROP FUNCTION f1;
