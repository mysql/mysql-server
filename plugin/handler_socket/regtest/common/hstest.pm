
# vim:sw=2:ai

package hstest;

use DBI;
use Net::HandlerSocket;

our %conf = ();

sub get_conf_env {
  my ($key, $defval) = @_;
  return $ENV{$key} || $defval;
}

sub init_conf {
  $conf{host} = get_conf_env("MYHOST", "localhost");
  $conf{myport} = get_conf_env("MYPORT", 3306);
  $conf{dbname} = get_conf_env("MYDBNAME", "hstestdb");
  $conf{ssps} = get_conf_env("MYSSPS");
  $conf{user} = get_conf_env("MYSQLUSER", "root");
  $conf{pass} = get_conf_env("MYSQLPASS", "");
  $conf{hsport} = get_conf_env("HSPORT", 9998);
  $conf{hspass} = get_conf_env("HSPASS", undef);
}

sub get_dbi_connection {
  my ($dbname, $host, $myport, $ssps, $user, $pass)
    = ($conf{dbname}, $conf{host}, $conf{myport}, $conf{ssps},
      $conf{user}, $conf{pass});
  my $mycnf = "binary_my.cnf";
  my $dsn = "DBI:mysql:database=;host=$host;port=$myport"
    . ";mysql_server_prepare=$ssps"
    . ";mysql_read_default_group=perl"
    . ";mysql_read_default_file=../common/$mycnf";
  my $dbh = DBI->connect($dsn, $user, $pass, { RaiseError => 1 });
  return $dbh;
}

sub init_testdb {
  my $charset = $_[0] || "binary";
  my $dbh = get_dbi_connection();
  my $dbname = $conf{dbname};
  $dbh->do("drop database if exists $dbname");
  $dbh->do("create database $dbname default character set $charset");
  $dbh->do("use $dbname");
  return $dbh;
}

sub get_hs_connection {
  my ($host, $port) = @_;
  $host ||= $conf{host};
  $port ||= $conf{hsport};
  my $hsargs = { 'host' => $host, 'port' => $port };
  my $conn = new Net::HandlerSocket($hsargs);
  if (defined($conn) && defined($conf{hspass})) {
    $conn->auth($conf{hspass});
  }
  return $conn;
}


init_conf();

1;

