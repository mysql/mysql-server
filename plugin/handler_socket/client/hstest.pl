#!/usr/bin/perl

# vim:sw=8:ai:ts=8

use strict;
use warnings;

use DBI;
use Net::HandlerSocket;

my %conf = ();
for my $i (@ARGV) {
	my ($k, $v) = split(/=/, $i);
	$conf{$k} = $v;
}

my $verbose = get_conf("verbose", 0);
my $actions_str = get_conf("actions", "hsread");
my $tablesize = get_conf("tablesize", 10000);
my $db = get_conf("db", "hstest");
my $table = get_conf("table", "hstest_table1");
my $engine = get_conf("engine", "innodb");
my $host = get_conf("host", "localhost");
my $mysqlport = get_conf("mysqlport", 3306);
my $mysqluser = get_conf("mysqluser", "root");
my $mysqlpass = get_conf("mysqlpass", "");
my $hsport = get_conf("hsport", 9999);
my $loop = get_conf("loop", 10000);
my $op = get_conf("op", "=");
my $ssps = get_conf("ssps", 0);
my $num_moreflds = get_conf("moreflds", 0);
my $moreflds_prefix = get_conf("moreflds_prefix", "column0123456789_");
my $keytype = get_conf("keytype", "varchar(32)");
my $file = get_conf("file", undef);

my $dsn = "DBI:mysql:database=;host=$host;port=$mysqlport"
	. ";mysql_server_prepare=$ssps";
my $dbh = DBI->connect($dsn, $mysqluser, $mysqlpass, { RaiseError => 1 });
my $hsargs = { 'host' => $host, 'port' => $hsport };
my $cli = new Net::HandlerSocket($hsargs);

my @actions = split(/,/, $actions_str);
for my $action (@actions) {
	if ($action eq "table") {
		print("TABLE $db.$table\n");
		$dbh->do("drop database if exists $db");
		$dbh->do("create database $db");
		$dbh->do("use $db");
		my $moreflds = get_createtbl_moreflds_str();
		$dbh->do(
			"create table $table (" .
			"k $keytype primary key" .
			",v varchar(32) not null" .
			$moreflds .
			") character set utf8 collate utf8_bin " .
			"engine = $engine");
	} elsif ($action eq "insert") {
		print("INSERT $db.$table tablesize=$tablesize\n");
		$dbh->do("use $db");
		my $moreflds = get_insert_moreflds_str();
		for (my $i = 0; $i < $tablesize; $i += 100) {
			my $qstr = "insert into $db.$table values";
			for (my $j = 0; $j < 100; ++$j) {
				if ($j != 0) {
					$qstr .= ",";
				}
				my $k = "" . ($i + $j);
				my $v = "v" . int(rand(1000)) . ($i + $j);
				$qstr .= "('$k', '$v'";
				for (my $j = 0; $j < $num_moreflds; ++$j) {
					$qstr .= ",'$j'";
				}
				$qstr .= ")";
			}
			$dbh->do($qstr);
			print "$i/$tablesize\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "read") {
		print("READ $db.$table op=$op loop=$loop\n");
		$dbh->do("use $db");
		my $moreflds = get_select_moreflds_str();
		my $sth = $dbh->prepare(
			"select k,v$moreflds from $db.$table where k = ?");
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			# print "k=$k\n";
			$sth->execute($k);
			if ($verbose >= 10) {
				print "RET:";
				while (my $ref = $sth->fetchrow_arrayref()) {
					my $rk = $ref->[0];
					my $rv = $ref->[1];
					print " $rk $rv";
				}
				print "\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsinsert") {
		print("HSINSERT $db.$table tablesize=$tablesize\n");
		$cli->open_index(1, $db, $table, '', 'k,v');
		for (my $i = 0; $i < $tablesize; ++$i) {
			my $k = "" . $i;
			my $v = "v" . int(rand(1000)) . $i;
			my $r = $cli->execute_insert(1, [ $k, $v ]);
			if ($r->[0] != 0) {
				die;
			}
			print "$i/$tablesize\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsread") {
		print("HSREAD $db.$table op=$op loop=$loop\n");
		my $moreflds = get_select_moreflds_str();
		$cli->open_index(1, $db, $table, '', "k,v$moreflds");
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			# print "k=$k\n";
			my $r = $cli->execute_find(1, $op, [ $k ], 1, 0);
			if ($verbose >= 10) {
				my $len = scalar(@{$r});
				print "LEN=$len";
				for my $e (@{$r}) {
					print " [$e]";
				}
				print "\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsupdate") {
		my $vbase = "v" . int(rand(1000));
		print("HSUPDATE $db.$table op=$op loop=$loop vbase=$vbase\n");
		$cli->open_index(1, $db, $table, '', 'v');
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			my $v = $vbase . $i;
			print "k=$k v=$v\n";
			my $r = $cli->execute_update(1, $op, [ $k ], 1, 0,
				[ $v ]);
			if ($verbose >= 10) {
				print "UP k=$k v=$v\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsdelete") {
		print("HSDELETE $db.$table op=$op loop=$loop\n");
		$cli->open_index(1, $db, $table, '', '');
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			print "k=$k\n";
			my $r = $cli->execute_delete(1, $op, [ $k ], 1, 0);
			if ($verbose >= 10) {
				print "DEL k=$k\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "verify") {
		verify_do();
	}
}

sub verify_do {
	my ($fail_cnt, $ok_cnt) = (0, 0);
	my $sth = $dbh->prepare("select v from $db.$table where k = ?");
	use FileHandle;
	my $fh = new FileHandle($file, "r");
	while (my $line = <$fh>) {
		chomp($line);
		my @vec = split(/\t/, $line);
		my $k = $vec[3];
		my $v = $vec[7];
		next if (!defined($k) || !defined($v));
		# print "$k $v\n";
		$sth->execute($k);
		my $aref = $sth->fetchrow_arrayref();
		if (!defined($aref)) {
			print "FAILED: $k notfound\n";
			++$fail_cnt;
		} else {
			my $gv = $aref->[0];
			if ($gv ne $v) {
				print "FAILED: $k got=$gv expected=$v\n";
				++$fail_cnt;
			} else {
				print "OK: $k $v $gv\n" if $verbose >= 10;
				++$ok_cnt;
			}
		}
	}
	print "OK=$ok_cnt FAIL=$fail_cnt\n";
}

sub get_conf {
	my ($key, $def) = @_;
	my $val = $conf{$key};
	if ($val) {
		print "$key=$val\n";
	} else {
		$val = $def;
		$def ||= '';
		print "$key=$def(default)\n";
	}
	return $val;
}

sub get_createtbl_moreflds_str {
	my $s = "";
	for (my $j = 0; $j < $num_moreflds; ++$j) {
		$s .= ",$moreflds_prefix$j varchar(30)";
	}
	return $s;
}

sub get_select_moreflds_str {
	my $s = "";
	for (my $i = 0; $i < $num_moreflds; ++$i) {
		$s .= ",$moreflds_prefix$i";
	}
	return $s;
}

sub get_insert_moreflds_str {
	my $s = "";
	for (my $i = 0; $i < $num_moreflds; ++$i) {
		$s .= ",?";
	}
	return $s;
}

