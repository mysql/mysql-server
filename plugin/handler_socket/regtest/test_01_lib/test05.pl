#!/usr/bin/perl

# vim:sw=2:ai

# test for binary cleanness (#3)

BEGIN {
	push @INC, "../common/";
};

use strict;
use warnings;
use hstest;

my $dbh = hstest::init_testdb();
my $table = 'hstesttbl';
my $tablesize = 256;
$dbh->do(
  "create table $table (k varchar(30) primary key, v varchar(30)) " .
  "engine = innodb default charset = binary");
srand(999);

my %valmap = ();

print "WR\n";
my $sth = $dbh->prepare("insert into $table values (?,?)");
for (my $i = 0; $i < $tablesize; ++$i) {
  my $k = "" . $i;
  my $v = ($i % 2 == 1) ? $i : undef;
  $sth->execute($k, $v);
  $v = "[null]" if !defined($v);
  print "$k $v\n";
  $valmap{$k} = $v;
}

my $hs = hstest::get_hs_connection();
my $dbname = $hstest::conf{dbname};
$hs->open_index(1, $dbname, $table, '', 'k,v');
my $r = $hs->execute_single(1, '>=', [ '' ], 10000, 0);
shift(@$r);
print "HS\n";
for (my $i = 0; $i < $tablesize; ++$i) {
  my $k = $r->[$i * 2];
  my $v = $r->[$i * 2 + 1];
  $v = "[null]" if !defined($v);
  print "$k $v\n";
  print "MISMATCH\n" if ($valmap{$k} ne $v);
}
undef $hs;

print "MY\n";
my $aref = $dbh->selectall_arrayref("select k,v from $table order by k");
for my $row (@$aref) {
  my ($k, $v) = @$row;
  $v = "[null]" if !defined($v);
  print "$k $v\n";
  print "MISMATCH\n" if ($valmap{$k} ne $v);
}

