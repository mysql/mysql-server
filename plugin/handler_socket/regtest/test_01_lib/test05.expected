WR
0 [null]
1 1
2 [null]
3 3
4 [null]
5 5
6 [null]
7 7
8 [null]
9 9
10 [null]
11 11
12 [null]
13 13
14 [null]
15 15
16 [null]
17 17
18 [null]
19 19
20 [null]
21 21
22 [null]
23 23
24 [null]
25 25
26 [null]
27 27
28 [null]
29 29
30 [null]
31 31
32 [null]
33 33
34 [null]
35 35
36 [null]
37 37
38 [null]
39 39
40 [null]
41 41
42 [null]
43 43
44 [null]
45 45
46 [null]
47 47
48 [null]
49 49
50 [null]
51 51
52 [null]
53 53
54 [null]
55 55
56 [null]
57 57
58 [null]
59 59
60 [null]
61 61
62 [null]
63 63
64 [null]
65 65
66 [null]
67 67
68 [null]
69 69
70 [null]
71 71
72 [null]
73 73
74 [null]
75 75
76 [null]
77 77
78 [null]
79 79
80 [null]
81 81
82 [null]
83 83
84 [null]
85 85
86 [null]
87 87
88 [null]
89 89
90 [null]
91 91
92 [null]
93 93
94 [null]
95 95
96 [null]
97 97
98 [null]
99 99
100 [null]
101 101
102 [null]
103 103
104 [null]
105 105
106 [null]
107 107
108 [null]
109 109
110 [null]
111 111
112 [null]
113 113
114 [null]
115 115
116 [null]
117 117
118 [null]
119 119
120 [null]
121 121
122 [null]
123 123
124 [null]
125 125
126 [null]
127 127
128 [null]
129 129
130 [null]
131 131
132 [null]
133 133
134 [null]
135 135
136 [null]
137 137
138 [null]
139 139
140 [null]
141 141
142 [null]
143 143
144 [null]
145 145
146 [null]
147 147
148 [null]
149 149
150 [null]
151 151
152 [null]
153 153
154 [null]
155 155
156 [null]
157 157
158 [null]
159 159
160 [null]
161 161
162 [null]
163 163
164 [null]
165 165
166 [null]
167 167
168 [null]
169 169
170 [null]
171 171
172 [null]
173 173
174 [null]
175 175
176 [null]
177 177
178 [null]
179 179
180 [null]
181 181
182 [null]
183 183
184 [null]
185 185
186 [null]
187 187
188 [null]
189 189
190 [null]
191 191
192 [null]
193 193
194 [null]
195 195
196 [null]
197 197
198 [null]
199 199
200 [null]
201 201
202 [null]
203 203
204 [null]
205 205
206 [null]
207 207
208 [null]
209 209
210 [null]
211 211
212 [null]
213 213
214 [null]
215 215
216 [null]
217 217
218 [null]
219 219
220 [null]
221 221
222 [null]
223 223
224 [null]
225 225
226 [null]
227 227
228 [null]
229 229
230 [null]
231 231
232 [null]
233 233
234 [null]
235 235
236 [null]
237 237
238 [null]
239 239
240 [null]
241 241
242 [null]
243 243
244 [null]
245 245
246 [null]
247 247
248 [null]
249 249
250 [null]
251 251
252 [null]
253 253
254 [null]
255 255
HS
0 [null]
1 1
10 [null]
100 [null]
101 101
102 [null]
103 103
104 [null]
105 105
106 [null]
107 107
108 [null]
109 109
11 11
110 [null]
111 111
112 [null]
113 113
114 [null]
115 115
116 [null]
117 117
118 [null]
119 119
12 [null]
120 [null]
121 121
122 [null]
123 123
124 [null]
125 125
126 [null]
127 127
128 [null]
129 129
13 13
130 [null]
131 131
132 [null]
133 133
134 [null]
135 135
136 [null]
137 137
138 [null]
139 139
14 [null]
140 [null]
141 141
142 [null]
143 143
144 [null]
145 145
146 [null]
147 147
148 [null]
149 149
15 15
150 [null]
151 151
152 [null]
153 153
154 [null]
155 155
156 [null]
157 157
158 [null]
159 159
16 [null]
160 [null]
161 161
162 [null]
163 163
164 [null]
165 165
166 [null]
167 167
168 [null]
169 169
17 17
170 [null]
171 171
172 [null]
173 173
174 [null]
175 175
176 [null]
177 177
178 [null]
179 179
18 [null]
180 [null]
181 181
182 [null]
183 183
184 [null]
185 185
186 [null]
187 187
188 [null]
189 189
19 19
190 [null]
191 191
192 [null]
193 193
194 [null]
195 195
196 [null]
197 197
198 [null]
199 199
2 [null]
20 [null]
200 [null]
201 201
202 [null]
203 203
204 [null]
205 205
206 [null]
207 207
208 [null]
209 209
21 21
210 [null]
211 211
212 [null]
213 213
214 [null]
215 215
216 [null]
217 217
218 [null]
219 219
22 [null]
220 [null]
221 221
222 [null]
223 223
224 [null]
225 225
226 [null]
227 227
228 [null]
229 229
23 23
230 [null]
231 231
232 [null]
233 233
234 [null]
235 235
236 [null]
237 237
238 [null]
239 239
24 [null]
240 [null]
241 241
242 [null]
243 243
244 [null]
245 245
246 [null]
247 247
248 [null]
249 249
25 25
250 [null]
251 251
252 [null]
253 253
254 [null]
255 255
26 [null]
27 27
28 [null]
29 29
3 3
30 [null]
31 31
32 [null]
33 33
34 [null]
35 35
36 [null]
37 37
38 [null]
39 39
4 [null]
40 [null]
41 41
42 [null]
43 43
44 [null]
45 45
46 [null]
47 47
48 [null]
49 49
5 5
50 [null]
51 51
52 [null]
53 53
54 [null]
55 55
56 [null]
57 57
58 [null]
59 59
6 [null]
60 [null]
61 61
62 [null]
63 63
64 [null]
65 65
66 [null]
67 67
68 [null]
69 69
7 7
70 [null]
71 71
72 [null]
73 73
74 [null]
75 75
76 [null]
77 77
78 [null]
79 79
8 [null]
80 [null]
81 81
82 [null]
83 83
84 [null]
85 85
86 [null]
87 87
88 [null]
89 89
9 9
90 [null]
91 91
92 [null]
93 93
94 [null]
95 95
96 [null]
97 97
98 [null]
99 99
MY
0 [null]
1 1
10 [null]
100 [null]
101 101
102 [null]
103 103
104 [null]
105 105
106 [null]
107 107
108 [null]
109 109
11 11
110 [null]
111 111
112 [null]
113 113
114 [null]
115 115
116 [null]
117 117
118 [null]
119 119
12 [null]
120 [null]
121 121
122 [null]
123 123
124 [null]
125 125
126 [null]
127 127
128 [null]
129 129
13 13
130 [null]
131 131
132 [null]
133 133
134 [null]
135 135
136 [null]
137 137
138 [null]
139 139
14 [null]
140 [null]
141 141
142 [null]
143 143
144 [null]
145 145
146 [null]
147 147
148 [null]
149 149
15 15
150 [null]
151 151
152 [null]
153 153
154 [null]
155 155
156 [null]
157 157
158 [null]
159 159
16 [null]
160 [null]
161 161
162 [null]
163 163
164 [null]
165 165
166 [null]
167 167
168 [null]
169 169
17 17
170 [null]
171 171
172 [null]
173 173
174 [null]
175 175
176 [null]
177 177
178 [null]
179 179
18 [null]
180 [null]
181 181
182 [null]
183 183
184 [null]
185 185
186 [null]
187 187
188 [null]
189 189
19 19
190 [null]
191 191
192 [null]
193 193
194 [null]
195 195
196 [null]
197 197
198 [null]
199 199
2 [null]
20 [null]
200 [null]
201 201
202 [null]
203 203
204 [null]
205 205
206 [null]
207 207
208 [null]
209 209
21 21
210 [null]
211 211
212 [null]
213 213
214 [null]
215 215
216 [null]
217 217
218 [null]
219 219
22 [null]
220 [null]
221 221
222 [null]
223 223
224 [null]
225 225
226 [null]
227 227
228 [null]
229 229
23 23
230 [null]
231 231
232 [null]
233 233
234 [null]
235 235
236 [null]
237 237
238 [null]
239 239
24 [null]
240 [null]
241 241
242 [null]
243 243
244 [null]
245 245
246 [null]
247 247
248 [null]
249 249
25 25
250 [null]
251 251
252 [null]
253 253
254 [null]
255 255
26 [null]
27 27
28 [null]
29 29
3 3
30 [null]
31 31
32 [null]
33 33
34 [null]
35 35
36 [null]
37 37
38 [null]
39 39
4 [null]
40 [null]
41 41
42 [null]
43 43
44 [null]
45 45
46 [null]
47 47
48 [null]
49 49
5 5
50 [null]
51 51
52 [null]
53 53
54 [null]
55 55
56 [null]
57 57
58 [null]
59 59
6 [null]
60 [null]
61 61
62 [null]
63 63
64 [null]
65 65
66 [null]
67 67
68 [null]
69 69
7 7
70 [null]
71 71
72 [null]
73 73
74 [null]
75 75
76 [null]
77 77
78 [null]
79 79
8 [null]
80 [null]
81 81
82 [null]
83 83
84 [null]
85 85
86 [null]
87 87
88 [null]
89 89
9 9
90 [null]
91 91
92 [null]
93 93
94 [null]
95 95
96 [null]
97 97
98 [null]
99 99
